File: backend/SkepticsTerminal.Core/Indicators/WickDetector.cs
using Skender.Stock.Indicators;
using SkepticsTerminal.Core.Models;

namespace SkepticsTerminal.Core.Indicators;

/// <summary>
/// Implements the "Ptil" (wick) detection logic.
/// An aggressive buy signal fires when the lower wick exceeds
/// 1.5× the average candle body, suggesting strong buying pressure at support.
/// </summary>
public static class WickDetector
{
    private const int LookbackPeriods = 20;

    /// <summary>
    /// Analyzes the most recent candle's wick structure.
    /// </summary>
    /// <param name="quotes">OHLCV history (oldest first). Uses last N candles.</param>
    /// <returns>Wick analysis for the most recent candle.</returns>
    public static WickAnalysis Analyze(IList<Quote> quotes)
    {
        if (quotes.Count < 2)
            return new WickAnalysis(0m, 0m, false);

        var current = quotes.Last();
        var window = quotes.TakeLast(LookbackPeriods).ToList();

        // Lower wick = distance from candle body bottom to the low
        var bodyBottom = Math.Min(current.Open, current.Close);
        var bodyTop = Math.Max(current.Open, current.Close);

        var lowerWick = bodyBottom - current.Low;
        var upperWick = current.High - bodyTop;

        // Average body size over lookback window (O(n) single pass)
        var avgBody = window
            .Select(q => (decimal)Math.Abs((double)(q.Close - q.Open)))
            .DefaultIfEmpty(0m)
            .Average();

        // Aggressive signal: lower wick is at least 1.5× the average body
        var hasAggressiveSignal = avgBody > 0m && lowerWick > avgBody * 1.5m;

        return new WickAnalysis(
            LowerWick: Math.Round(lowerWick, 2),
            UpperWick: Math.Round(upperWick, 2),
            HasAggressiveBuySignal: hasAggressiveSignal
        );
    }
}

File: api/lib/indicators/wick-detector.ts
import type { OhlcvBar, WickAnalysis } from '../types';

const LOOKBACK = 20;

/** Ptil wick logic: fires when lowerWick > avgBody * 1.5 */
export function detectWick(bars: OhlcvBar[]): WickAnalysis {
  if (bars.length < 2) return { lowerWick: 0, upperWick: 0, hasAggressiveBuySignal: false };

  const current = bars[bars.length - 1];
  const window = bars.slice(-LOOKBACK);

  const bodyBottom = Math.min(current.open, current.close);
  const bodyTop = Math.max(current.open, current.close);
  const lowerWick = parseFloat((bodyBottom - current.low).toFixed(2));
  const upperWick = parseFloat((current.high - bodyTop).toFixed(2));

  const avgBody = window.reduce((sum, b) => sum + Math.abs(b.close - b.open), 0) / window.length;
  const hasAggressiveBuySignal = avgBody > 0 && lowerWick > avgBody * 1.5;

  return { lowerWick, upperWick, hasAggressiveBuySignal };
}
